core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -942,8 +935,27 @@ private RetryPlan makeRetryDecision(Status status, Metadata trailer) {
         } // else no retry
       } // else no retry
 
-      return new RetryPlan(
-          shouldRetry, /* isFatal = */ false, backoffNanos, isHedging ? pushbackMillis : null);
+      return new RetryPlan(shouldRetry, backoffNanos);
+    }
+
+    private HedgingPlan makeHedgingDecision(Status status, Metadata trailer) {
+      String pushbackStr = trailer.get(GRPC_RETRY_PUSHBACK_MS);
+      Integer pushbackMillis = null;
+      if (pushbackStr != null) {
+        try {
+          pushbackMillis = Integer.valueOf(pushbackStr);
+        } catch (NumberFormatException e) {
+          pushbackMillis = -1;
+        }
+      }
+      boolean isFatal = !hedgingPolicy.nonFatalStatusCodes.contains(status.getCode());
+      boolean isThrottled = false;
+      if (throttle != null) {
+        if (!isFatal || (pushbackMillis != null && pushbackMillis < 0)) {
```

Comment:
Can this be `&&` instead of `||`? Or, even better, make line 953 `if (!isFatal && throttle != null)`? If `isFatal=true`, the return statement below will always be equivalent to `return new HedgingPlan(false, pushbackMillis)`.

Reply:
I don't quite follow. `throttle.onQualifiedFailure()` must be called if `!isFatal || (pushbackMillis != null && pushbackMillis < 0)` regardless what HedgingPlan will be returned.

Reply:
Ah, didn't realize that method had side effects